cdc/scheduler/agent.go

```diff
@@ -0,0 +1,448 @@
+// implements its own TableExecutor and ProcessorMessenger.
+type BaseAgent struct {
+	executor     TableExecutor
+	communicator ProcessorMessenger
+
+	// pendingOpsMu protects pendingOps.
+	// Notes that we need a mutex because some methods are expected
+	// to be called from a message handler goroutine.
+	pendingOpsMu sync.Mutex
+	// pendingOps is a queue of operations yet to be processed.
+	// the Deque stores *agentOperation.
+	pendingOps deque.Deque
+
+	// tableOperations is a map from tableID to the operation
+	// that is currently being processed.
+	tableOperations map[model.TableID]*agentOperation
+
+	// needSyncNow indicates that the agent needs to send the
+	// current owner a sync message as soon as possible.
+	needSyncNow *atomic.Bool
+
+	// checkpointSender is used to send checkpoint-ts to the owner.
+	checkpointSender checkpointSender
+
+	ownerInfoMu sync.RWMutex
+	ownerInfo   *ownerInfo
+
+	// ownerHasChanged indicates that the owner has changed and
+	// the communicator needs to be reset.
+	ownerHasChanged *atomic.Bool
+
+	// read-only fields
+	config *BaseAgentConfig
+	logger *zap.Logger
+}
+
+// NewBaseAgent creates a new BaseAgent.
+func NewBaseAgent(
+	changeFeedID model.ChangeFeedID,
+	executor TableExecutor,
+	messenger ProcessorMessenger,
+	config *BaseAgentConfig,
+) *BaseAgent {
+	logger := log.L().With(zap.String("changefeed-id", changeFeedID))
+	return &BaseAgent{
+		pendingOps:       deque.NewDeque(),
+		tableOperations:  map[model.TableID]*agentOperation{},
+		logger:           logger,
+		executor:         executor,
+		ownerInfo:        &ownerInfo{},
+		communicator:     messenger,
+		needSyncNow:      atomic.NewBool(true),
+		checkpointSender: newCheckpointSender(messenger, logger, config.SendCheckpointTsInterval),
+		ownerHasChanged:  atomic.NewBool(false),
+		config:           config,
+	}
+}
+
+type agentOperationStatus int32
+
+const (
+	operationReceived = agentOperationStatus(iota)
+	operationProcessed
+	operationFinished
+)
+
+type agentOperation struct {
+	TableID  model.TableID
+	IsDelete bool
+
+	status agentOperationStatus
+}
+
+type ownerInfo struct {
+	OwnerCaptureID model.CaptureID
+	// OwnerRev is needed in order to know who is the latest owner
+	// whenever there is a possibility of confusion, usually when the
+	// old owner has just gone down but its gRPC messages have yet to be
+	// processed. Since messages from the old and new owner could interleave,
+	// we need a way to tell.
+	OwnerRev int64
+}
+
+// Tick implements the interface Agent.
+func (a *BaseAgent) Tick(ctx context.Context) error {
+	if a.ownerHasChanged.Swap(false) {
+		// We need to notify the communicator if the owner has changed.
+		// This is necessary because the communicator might be waiting for
+		// messages to be received by the previous owner.
+		a.communicator.OnOwnerChanged(ctx, a.currentOwner())
+	}
+
+	if err := a.sendCheckpoint(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	if a.needSyncNow.Load() {
```

Comment:
Can processor send checkpoint before a needed `sync`?

Reply:
The Protocol does not forbid this behavior. There are two scenarios where we need `sync`:
- Processor initialization: Since we forbids the processor from sending checkpoints when no table is running, we cannot send checkpoint in this situation.
- In response to owner's announce: The checkpoint sent after receiving an announce would be the expected correct checkpoint, and the owner will take it into consideration if the checkpoint is received after `sync`.

Although there is no obvious correctness problem, I will modify the code a little bit so there will be less confusion.